design/src/rfcs/rfc0025_better_constraint_violations.md

```diff
@@ -0,0 +1,511 @@
+
+                    Ok((k, v))
+                })
+                .collect();
+            let hm = res?;
+            Self::try_from(hm)
+        }
+    }
+}
+```
+
+In conclusion, the user is currently exposed to an internal detail of how the
+framework operates that has no bearing on their application code. They
+shouldn't be exposed to impossible constraint violation variants in their Rust
+docs, nor have to `match` on these variants when handling errors.
+
+Note: [this comment] alludes to the problem described above.
+
+[this comment]: https://github.com/awslabs/smithy-rs/blob/27020be3421fb93e35692803f9a795f92feb1d19/codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/MapConstraintViolationGenerator.kt#L66-L69
+
+### Solution proposal
+
+The problem can be mitigated by adding `#[doc(hidden)]` to the internal
+variants and `#[non_exhaustive]` to the enum. We're already doing this in some
+constraint violation types.
+
+However, a "less leaky" solution is achieved by _splitting_ the constraint
+violation type into two types:
+
+1. one for use by the framework, with `pub(crate)` visibility, named
+   `ConstraintViolationException`; and
+2. one for use by user application code, with `pub` visibility, named
+   `ConstraintViolation`.
+
+```rust
+pub mod length_map {
+    pub enum ConstraintViolation {
+        Length(usize),
+    }
+    pub (crate) enum ConstraintViolationException {
+        Length(usize),
+        Value(
+            std::string::String,
+            crate::model::length_string::ConstraintViolation,
+        ),
+    }
+}
+```
+
+Note that, to some extent, the spirit of this approach is [already currently
+present](https://github.com/awslabs/smithy-rs/blob/9a4c1f304f6f5237d480cfb56dad2951d927d424/codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerBuilderGenerator.kt#L78-L81)
+in the case of builder types when `publicConstrainedTypes` is set to `false`:
+
+1. [`ServerBuilderGenerator.kt`] renders the usual builder type that enforces
+   constraint traits, setting its visibility to `pub (crate)`, for exclusive
+   use by the framework.
+2. [`ServerBuilderGeneratorWithoutPublicConstrainedTypes.kt`] renders the
+   builder type the user is exposed to: this builder does not take in
+   constrained types and does not enforce all modeled constraints.
+
+[`ServerBuilderGenerator.kt`]: https://github.com/awslabs/smithy-rs/blob/2226feff8f7fa884204f81a50d7e016386912acc/codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerBuilderGenerator.kt
+[`ServerBuilderGeneratorWithoutPublicConstrainedTypes.kt`]: https://github.com/awslabs/smithy-rs/blob/2226feff8f7fa884204f81a50d7e016386912acc/codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerBuilderGeneratorWithoutPublicConstrainedTypes.kt
+
+Collecting constraint violations
```

Comment:
I am not sure this title convey properly what you are explaining below, but I don't have a better name on top of my mind.